router/utils/utils.go

```diff
@@ -92,3 +100,60 @@ func EnhanceJsonWithTime(t time.Time, key string, resp []byte) []byte {
 func IsNotEmptyString(s string) bool {
 	return len(strings.TrimSpace(s)) > 0
 }
+
+type Drainer interface {
+	Drain(
+		job *jobsdb.JobT,
+		jobParams JobParameters,
+	) (bool, string)
+}
+
+func NewDrainer(
+	conf *config.Config,
+	destDrainFunc func(string) (bool, string),
+) Drainer {
+	return &drainer{
+		destinationIDs: conf.GetReloadableStringSliceVar(
+			[]string{},
+			"Router.toAbortDestinationIDs",
+		),
+		jobRunIDs: conf.GetReloadableStringSliceVar(
+			[]string{},
+			"RSources.toAbortJobRunIDs",
+		),
+		destDrainFunc: destDrainFunc,
+	}
+}
+
+type drainer struct {
+	destinationIDs misc.ValueLoader[[]string]
+	jobRunIDs      misc.ValueLoader[[]string]
+
+	destDrainFunc func(string) (bool, string)
+}
+
+func (d *drainer) Drain(
+	job *jobsdb.JobT,
+	jobParams JobParameters,
+) (bool, string) {
+	createdAt := job.CreatedAt
+	destID := jobParams.DestinationID
+	if time.Since(createdAt) > getRetentionTimeForDestination(destID) {
+		return true, "job expired"
+	}
+
+	if destDrain, destDrainReason := d.destDrainFunc(destID); destDrain {
+		return true, destDrainReason
+	}
+
+	if slices.Contains(d.destinationIDs.Load(), destID) {
+		return true, DestAbort
+	}
+
+	if jobParams.SourceJobRunID != "" &&
+		slices.Contains(d.jobRunIDs.Load(), jobParams.SourceJobRunID) {
+		return true, JobRunIDCancelled
+	}
+
+	return false, ""
+}
```

Comment:
This makes more sense to me, i.e. 
1. drainer can decide whether a job can be drained or not, as long as it resolve the destination's configuration from the destination id
2. The only argument `Drain` method should accept is `*jobsdb.JobT` 

```suggestion
type drainer struct {
	destinationIDs      misc.ValueLoader[[]string]
	jobRunIDs           misc.ValueLoader[[]string]
	destinationResolver func(string) (*backendconfig.DestinationT, bool)
}

func (d *drainer) Drain(job *jobsdb.JobT) (bool, string) {
	createdAt := job.CreatedAt
	var parameters JobParameters
	_ = json.Unmarshal(job.Parameters, &parameters)
	destID := parameters.DestinationID
	if time.Since(createdAt) > getRetentionTimeForDestination(destID) {
		return true, "job expired"
	}

	if destination, ok := d.destinationResolver(destID); !ok {
		return true, DestNotFoundInConfig
	} else if !destination.Enabled {
		return true, DestDisabled
	}

	if slices.Contains(d.destinationIDs.Load(), destID) {
		return true, DestAbort
	}

	if parameters.SourceJobRunID != "" &&
		slices.Contains(d.jobRunIDs.Load(), parameters.SourceJobRunID) {
		return true, JobRunIDCancelled
	}

	return false, ""
}
```